File: javascript/src/tuple/Empty.ts
//··························································
// Copyright (c) 2023-2026. Jonathan Bédard ~ JóôòKiwi
//
// This project is free to use.
// All the right is reserved to the author of this project.
// My projects:
//  - https://github.com/joooKiwi/type
//  - https://github.com/joooKiwi/lazy
//  - https://github.com/joooKiwi/collection
//  - https://github.com/joooKiwi/enumeration
//··························································

import type {EmptyArray, EmptySet} from "@joookiwi/type"

import type {Tuple}     from "./Tuple"
import type {TupleName} from "../type/toStringTag"

import {CollectionConstants}     from "../CollectionConstants"
import {EmptyCollectionHolder}   from "../EmptyCollectionHolder"
import {EmptyCollectionIterator} from "../iterator/EmptyCollectionIterator"

export class Empty
    implements Tuple<never> {

    //#region -------------------- Singleton usage --------------------

    static #instance?: Empty

    protected constructor() {}

    public static get get(): Empty {
        return Empty.#instance ??= new Empty()
    }

    //#endregion -------------------- Singleton usage --------------------

    //#region -------------------- Size methods --------------------

    public get size(): 0 { return 0 }
    public get length(): this["size"] { return this.size }
    public get count(): this["size"] { return this.size }

    public get isEmpty(): true { return true }
    public get isNotEmpty(): false { return false }

    public get hasAtLeast1Element(): this["isNotEmpty"] { return this.isNotEmpty }
    public get includesAtLeast1Element(): this["isNotEmpty"] { return this.isNotEmpty }
    public get containsAtLeast1Element(): this["isNotEmpty"] { return this.isNotEmpty }

    public get hasExactly1Element(): false { return false }
    public get includesExactly1Element(): this["hasExactly1Element"] { return this.hasExactly1Element }
    public get containsExactly1Element(): this["hasExactly1Element"] { return this.hasExactly1Element }

    public get hasAtMost1Element(): true { return true }
    public get includesAtMost1Element(): this["hasAtMost1Element"] { return this.hasAtMost1Element }
    public get containsAtMost1Element(): this["hasAtMost1Element"] { return this.hasAtMost1Element }

    public get hasAtLeast2Elements(): false { return false }
    public get includesAtLeast2Elements(): this["hasAtLeast2Elements"] { return this.hasAtLeast2Elements }
    public get containsAtLeast2Elements(): this["hasAtLeast2Elements"] { return this.hasAtLeast2Elements }

    public get hasExactly2Elements(): false { return false }
    public get includesExactly2Elements(): this["hasExactly2Elements"] { return this.hasExactly2Elements }
    public get containsExactly2Elements(): this["hasExactly2Elements"] { return this.hasExactly2Elements }

    public get hasAtMost2Elements(): true { return true }
    public get includesAtMost2Elements(): this["hasAtMost2Elements"] { return this.hasAtMost2Elements }
    public get containsAtMost2Elements(): this["hasAtMost2Elements"] { return this.hasAtMost2Elements }

    //#endregion -------------------- Size methods --------------------
    //#region -------------------- Has ‥ --------------------

    public get hasNull(): false { return false }
    public get includesNull(): this["hasNull"] { return this.hasNull }
    public get containsNull(): this["hasNull"] { return this.hasNull }

    public get hasNoNulls(): true { return true }
    public get includesNoNulls(): this["hasNoNulls"] { return this.hasNoNulls }
    public get containsNoNulls(): this["hasNoNulls"] { return this.hasNoNulls }

    public get hasDuplicate(): false { return false }
    public get includesDuplicate(): this["hasDuplicate"] { return this.hasDuplicate }
    public get containsDuplicate(): this["hasDuplicate"] { return this.hasDuplicate }

    public get hasNoDuplicates(): true { return true }
    public get includesNoDuplicates(): this["hasNoDuplicates"] { return this.hasNoDuplicates }
    public get containsNoDuplicates(): this["hasNoDuplicates"] { return this.hasNoDuplicates }

    //#endregion -------------------- Has ‥ --------------------
    //#region -------------------- JavaScript methods --------------------

    public get [Symbol.toStringTag](): TupleName { return "Tuple" }

    //#endregion -------------------- JavaScript methods --------------------
    //#region -------------------- Conversion methods --------------------

    public toIterator(): EmptyCollectionIterator { return EmptyCollectionIterator.get }

    public toArray(): EmptyArray { return CollectionConstants.EMPTY_ARRAY }

    public toSet(): EmptySet { return CollectionConstants.EMPTY_SET }

    public toCollection(): EmptyCollectionHolder { return EmptyCollectionHolder.get }

    public toString(): string { return "Empty ()" }

    //#endregion -------------------- Conversion methods --------------------

}

File: java/collection.joookiwi/src/main/java/joookiwi/collection/java/tuple/Empty.java
package joookiwi.collection.java.tuple;

import joookiwi.collection.java.EmptyCollectionHolder;
import joookiwi.collection.java.annotation.Alias;
import joookiwi.collection.java.annotation.Singleton;
import joookiwi.collection.java.method.ArrayCreator;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNullByDefault;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.Range;

import static joookiwi.collection.java.CommonContracts.ALWAYS_FALSE_0;
import static joookiwi.collection.java.CommonContracts.ALWAYS_FALSE_1;
import static joookiwi.collection.java.CommonContracts.ALWAYS_TRUE_0;
import static joookiwi.collection.java.CommonContracts.ALWAYS_TRUE_1;
import static joookiwi.collection.java.CommonContracts.IF_1ST_NULL_THEN_FALSE_1;
import static joookiwi.collection.java.CommonContracts._1ST_IS_NOT_NULL_1;
import static joookiwi.collection.java.annotation.AliasArgument.IGNORED;

/// A [Tuple] with no values (as a [singleton][Singleton] instance)
///
/// @param <T> The type
/// @see Monuple
/// @see Couple
@Singleton
@NotNullByDefault
public class Empty<T extends @Nullable Object>
        implements Tuple<T> {

    //#region -------------------- Singleton usage --------------------

    private static volatile @Nullable Empty<?> instance;

    protected Empty() { super(); }

    @SuppressWarnings("unchecked cast")
    public static <T> Empty<T> getInstance() {
        final var instance = Empty.instance;
        if (instance != null)
            return (Empty<T>) instance;
        synchronized (Empty.class) {
            final var instance2 = Empty.instance;
            if (instance2 != null)
                return (Empty<T>) instance2;
            return (Empty<T>) (Empty.instance = new Empty<>());
        }
    }

    //#endregion -------------------- Singleton usage --------------------

    //#region -------------------- Size methods --------------------

    @Contract(pure = true) @Override public @Range(from = 0, to = 0) int size() { return 0; }
    @Contract(pure = true) @Override public @Range(from = 0, to = 0) final int length() { return size(); }
    @Contract(pure = true) @Override public @Range(from = 0, to = 0) final int count() { return size(); }


    @Contract(value = ALWAYS_TRUE_0, pure = true) @Override public boolean isEmpty() { return true; }
    @Contract(value = ALWAYS_FALSE_0, pure = true) @Override public boolean isNotEmpty() { return false; }


    @Contract(value = ALWAYS_FALSE_0, pure = true) @Override public final boolean hasAtLeast1Element() { return isNotEmpty(); }
    @Contract(value = ALWAYS_FALSE_0, pure = true) @Override public final boolean containsAtLeast1Element() { return isNotEmpty(); }
    @Contract(value = ALWAYS_FALSE_0, pure = true) @Override public final boolean includesAtLeast1Element() { return isNotEmpty(); }

    @Contract(value = ALWAYS_FALSE_0, pure = true) @Override public boolean hasExactly1Element() { return false; }
    @Contract(value = ALWAYS_FALSE_0, pure = true) @Override public final boolean containsExactly1Element() { return hasExactly1Element(); }
    @Contract(value = ALWAYS_FALSE_0, pure = true) @Override public final boolean includesExactly1Element() { return hasExactly1Element(); }

    @Contract(value = ALWAYS_TRUE_0, pure = true) @Override public boolean hasAtMost1Element() { return true; }
    @Contract(value = ALWAYS_TRUE_0, pure = true) @Override public final boolean containsAtMost1Element() { return hasAtMost1Element(); }
    @Contract(value = ALWAYS_TRUE_0, pure = true) @Override public final boolean includesAtMost1Element() { return hasAtMost1Element(); }


    @Contract(value = ALWAYS_FALSE_0, pure = true) @Override public boolean hasAtLeast2Elements() { return false; }
    @Contract(value = ALWAYS_FALSE_0, pure = true) @Override public final boolean containsAtLeast2Elements() { return hasAtLeast2Elements(); }
    @Contract(value = ALWAYS_FALSE_0, pure = true) @Override public final boolean includesAtLeast2Elements() { return hasAtLeast2Elements(); }

    @Contract(value = ALWAYS_FALSE_0, pure = true) @Override public boolean hasExactly2Elements() { return false; }
    @Contract(value = ALWAYS_FALSE_0, pure = true) @Override public final boolean containsExactly2Elements() { return hasExactly2Elements(); }
    @Contract(value = ALWAYS_FALSE_0, pure = true) @Override public final boolean includesExactly2Elements() { return hasExactly2Elements(); }

    @Contract(value = ALWAYS_TRUE_0, pure = true) @Override public boolean hasAtMost2Elements() { return true; }
    @Contract(value = ALWAYS_TRUE_0, pure = true) @Override public final boolean containsAtMost2Elements() { return hasAtMost2Elements(); }
    @Contract(value = ALWAYS_TRUE_0, pure = true) @Override public final boolean includesAtMost2Elements() { return hasAtMost2Elements(); }

    //#endregion -------------------- Size methods --------------------
    //#region -------------------- Has ‥ --------------------

    //#region -------------------- Has null --------------------

    @Contract(ALWAYS_FALSE_0) @Override public boolean hasNull(                                            ) { return false; }
    @Contract(ALWAYS_FALSE_1)           public boolean hasNull(final @Nullable Object @Nullable ... ignored) { return false; }

    @Contract(ALWAYS_FALSE_0)                                                @Override public final boolean includesNull(                                            ) { return hasNull(); }
    @Contract(ALWAYS_FALSE_1) @Alias(value = "hasNull", arguments = IGNORED)           public final boolean includesNull(final @Nullable Object @Nullable ... ignored) { return hasNull(); }

    @Contract(ALWAYS_FALSE_0)                                                @Override public final boolean containsNull(                                            ) { return hasNull(); }
    @Contract(ALWAYS_FALSE_1) @Alias(value = "hasNull", arguments = IGNORED)           public final boolean containsNull(final @Nullable Object @Nullable ... ignored) { return hasNull(); }

    //#endregion -------------------- Has null --------------------
    //#region -------------------- Has no nulls --------------------

    @Contract(ALWAYS_TRUE_0) @Override public boolean hasNoNulls(                                            ) { return true; }
    @Contract(ALWAYS_TRUE_1)           public boolean hasNoNulls(final @Nullable Object @Nullable ... ignored) { return true; }

    @Contract(ALWAYS_TRUE_0)                                                   @Override public final boolean includesNoNulls(                                            ) { return hasNoNulls(); }
    @Contract(ALWAYS_TRUE_1) @Alias(value = "hasNoNulls", arguments = IGNORED)           public final boolean includesNoNulls(final @Nullable Object @Nullable ... ignored) { return hasNoNulls(); }

    @Contract(ALWAYS_TRUE_0)                                                   @Override public final boolean containsNoNulls(                                            ) { return hasNoNulls(); }
    @Contract(ALWAYS_TRUE_1) @Alias(value = "hasNoNulls", arguments = IGNORED)           public final boolean containsNoNulls(final @Nullable Object @Nullable ... ignored) { return hasNoNulls(); }

    //#endregion -------------------- Has no nulls --------------------
    //#region -------------------- Has duplicate --------------------

    @Contract(ALWAYS_FALSE_0) @Override public boolean hasDuplicate(                                            ) { return false; }
    @Contract(ALWAYS_FALSE_1)           public boolean hasDuplicate(final @Nullable Object @Nullable ... ignored) { return false; }

    @Contract(ALWAYS_FALSE_0)                                                     @Override public final boolean includesDuplicate(                                            ) { return hasDuplicate(); }
    @Contract(ALWAYS_FALSE_1) @Alias(value = "hasDuplicate", arguments = IGNORED)           public final boolean includesDuplicate(final @Nullable Object @Nullable ... ignored) { return hasDuplicate(); }

    @Contract(ALWAYS_FALSE_0)                                                     @Override public final boolean containsDuplicate(                                            ) { return hasDuplicate(); }
    @Contract(ALWAYS_FALSE_1) @Alias(value = "hasDuplicate", arguments = IGNORED)           public final boolean containsDuplicate(final @Nullable Object @Nullable ... ignored) { return hasDuplicate(); }

    //#endregion -------------------- Has duplicate --------------------
    //#region -------------------- Has no duplicates --------------------

    @Contract(ALWAYS_TRUE_0) @Override public boolean hasNoDuplicates(                                            ) { return true; }
    @Contract(ALWAYS_TRUE_1)           public boolean hasNoDuplicates(final @Nullable Object @Nullable ... ignored) { return true; }

    @Contract(ALWAYS_TRUE_0)                                                        @Override public final boolean includesNoDuplicates(                                            ) { return hasNoDuplicates(); }
    @Contract(ALWAYS_TRUE_1) @Alias(value = "hasNoDuplicates", arguments = IGNORED)           public final boolean includesNoDuplicates(final @Nullable Object @Nullable ... ignored) { return hasNoDuplicates(); }

    @Contract(ALWAYS_TRUE_0)                                                        @Override public final boolean containsNoDuplicates(                                            ) { return hasNoDuplicates(); }
    @Contract(ALWAYS_TRUE_1) @Alias(value = "hasNoDuplicates", arguments = IGNORED)           public final boolean containsNoDuplicates(final @Nullable Object @Nullable ... ignored) { return hasNoDuplicates(); }

    //#endregion -------------------- Has no duplicates --------------------

    //#endregion -------------------- Has ‥ --------------------
    //#region -------------------- Conversion methods --------------------

    @Override public T[] toArray() { return ArrayCreator.Array(); }

    @Override public EmptyCollectionHolder<T> toCollection() { return EmptyCollectionHolder.getInstance(); }

    @Override public String toString() { return "Empty ()"; }

    //#endregion -------------------- Conversion methods --------------------
    //#region -------------------- Comparison methods --------------------

    @Contract(pure = true) @Override public @Range(from = 0, to = 0) int hashCode() { return 0; }

    //#region -------------------- Equals --------------------

    @Contract(value = IF_1ST_NULL_THEN_FALSE_1, pure = true)
    @Override public boolean equals(final @Nullable Object other) {
        if (other == null)
            return false;
        if (other == this)
            return true;
        if (other instanceof Empty<?> otherConverted)
            return otherConverted.isEmpty();
        return false;
    }

    @Contract(value = IF_1ST_NULL_THEN_FALSE_1, pure = true)
    public boolean equals(final @Nullable Tuple<?> other) {
        if (other == null)
            return false;
        if (other == this)
            return true;
        return other.isEmpty();
    }

    @Contract(value = _1ST_IS_NOT_NULL_1, pure = true) public boolean equals(final @Nullable Empty<?> other) { return other != null; }

    //#endregion -------------------- Equals --------------------
    //#region -------------------- Reference equals --------------------

    @Contract(value = IF_1ST_NULL_THEN_FALSE_1, pure = true) @Override public boolean referenceEquals(final @Nullable Object           other) { return other == this; }
    @Contract(value = IF_1ST_NULL_THEN_FALSE_1, pure = true)           public boolean referenceEquals(final @Nullable Tuple<?>         other) { return other == this; }
    @Contract(value = ALWAYS_FALSE_1,           pure = true)           public boolean referenceEquals(final @Nullable AbstractTuple<?> other) { return false; }
    @Contract(value = IF_1ST_NULL_THEN_FALSE_1, pure = true)           public boolean referenceEquals(final @Nullable Empty<?>         other) { return other == this; }

    //#endregion -------------------- Reference equals --------------------

    //#endregion -------------------- Comparison methods --------------------
    //#region -------------------- Clone methods --------------------

    @Override public Empty<T> clone() { return this; }

    //#endregion -------------------- Clone methods --------------------

}
